Makefile

```diff
@@ -154,6 +154,64 @@ test-conformance: $(BIN)/conformance_test_runner $(BUILD)/protobuf-conformance
 		--text_format_failure_list packages/protobuf-conformance/conformance_failing_tests_text_format.txt \
 		packages/protobuf-conformance/bin/conformance_esm.js
 
+TS_VERSIONS = 4.1.2 \
+			  4.1.3 \
+			  4.1.4 \
+			  4.1.5 \
+			  4.1.6 \
+			  4.2.2 \
+			  4.2.3 \
+			  4.2.4 \
+			  4.3.2 \
+			  4.3.3 \
+			  4.3.4 \
+			  4.3.5 \
+			  4.4.2 \
+			  4.4.3 \
+			  4.4.4 \
+			  4.5.2 \
+			  4.5.3 \
+			  4.5.4 \
+			  4.5.5 \
+			  4.6.2 \
+			  4.6.3 \
+			  4.6.4 \
+			  4.7.2 \
+			  4.7.3 \
+			  4.7.4 \
+			  4.8.1-rc \
+
+.PHONY: install-ts
+install-ts:  node_modules
+	@for number in $(TS_VERSIONS) ; do \
+		formatted=$$(echo "$${number}" | sed -r 's/[\.]/_/g'); \
+		dirname=packages/protobuf-test/typescript/ts$${formatted} ; \
+		if [ ! -f $$dirname/tsconfig.json ]; then \
+			npm i -w packages/protobuf-test ts$${formatted}@npm:typescript@$${number}; \
+			echo "Installing TypeScript `node_modules/ts$$formatted/bin/tsc --version`" ; \
+			mkdir -p $$dirname/ ; \
+			node_modules/ts$$formatted/bin/tsc --init ; \
+			mv tsconfig.json $$dirname/tsconfig.json ; \
+		else \
+	    	echo "Version $$number already installed" ; \
+		fi ; \
+	done
+
+.PHONY: test-ts-compat
+test-ts-compat: $(GEN)/protobuf-test node_modules $(shell find packages/protobuf-test -name '*.json')
+	@for number in $(TS_VERSIONS) ; do \
+		formatted=$$(echo "$${number}" | sed -r 's/[\.]/_/g'); \
+		dirname=packages/protobuf-test/typescript/ts$${formatted} ; \
+		echo "Testing TypeScript `node_modules/ts$$formatted/bin/tsc --version`" ; \
+		node_modules/ts$$formatted/bin/tsc -p $$dirname/tsconfig.json --noEmit > $$dirname/errors.log 2>&1; \
+		if [ "$$?" -eq 0 ]; then \
+	  		echo "Success" ; \
+			rm $$dirname/errors.log ; \
+  		else \
+	  		echo "Failed compilation.  Errors saved to $${dirname}/errors.log" ; \
```

Comment:
I see your point in redirecting the error output, but I'd prefer if we can keep this as simple as possible, especially in the Makefile. Failing early seems totally fine to me.

Reply:
Sounds good.  I kinda like redirecting the error output for two reasons:

1.  It's easier to debug errors that may occur rather than spewing them to the terminal (esp. if there's a lot)
2. Failing early means that any TS versions not tested could also be failing (and for different reasons) but you won't know until you fix the first error.

My preference would be to test all versions, write to a log, and then exit if failure.  But def not a huge deal to me.  I fixed this so that it just exits on the first failure and doesn't write to any log files.